### PR TITLE
Web3Auth, Take 2

### PR DESCRIPTION
**Pull request by Arshia001**
This PR includes the following:

* An in-process signer and transaction executor
* A new UPM package for web3auth functionality
* A web3auth wallet that handles logging in and uses the in-process stuff internally after the private key is retrieved from W3A
* Sample code for using the web3auth wallet functionality to log in. Note: no new sample code is required outside of logging in since the implementation is fully compatible with the core interfaces.

This code was created in a hurry, so it should be tested extensively. Also, please make sure the `SignTypedData` functionality is working correctly with the Gelato module.

Also, most of the 188 changed files are renames, the PR contains little new code.

### CODE REVIEW

**Comment by Arshia001** [comment]
@juans-chainsafe To test this PR, you need to log in via W3A in the login scene, and then make sure all the prefabs are working correctly as they were before. To be able to test transactions, you'd need to send funds to the wallet that is generated by W3A. You would probably need to export the private key and import it into MetaMask to be able to do that. Right now, there is no way to export the private key.

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
@Arshia001 I didn't finish my testing, but so far these are some of the issues that I found after I did the login with google (I have funds in my address)

Get Gas Limit
```
Web3Exception: RPC returned error for "eth_estimateGas": -32000 execution reverted 
Web3Unity.Scripts.Library.Ethers.Providers.JsonRpcProvider.Perform[T] (System.String method, System.Object[] parameters) (at <5ae973ee2bcf42eb922496935c4c57aa>:0)
Rethrow as Web3Exception: eth_estimateGas: bad result from RPC endpoint
Web3Unity.Scripts.Library.Ethers.Providers.JsonRpcProvider.Perform[T] (System.String method, System.Object[] parameters) (at <5ae973ee2bcf42eb922496935c4c57aa>:0)
Web3Unity.Scripts.Library.Ethers.Providers.RpcProviderExtensions.EstimateGas (Web3Unity.Scripts.Library.Ethers.Providers.IRpcProvider provider, Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <5ae973ee2bcf42eb922496935c4c57aa>:0)
Web3Unity.Scripts.Library.Ethers.Contracts.Contract.EstimateGas (System.String method, System.Object[] parameters, Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest overwrite) (at <5ae973ee2bcf42eb922496935c4c57aa>:0)
GetGasLimit.Start () (at Assets/Scripts/Samples/Wallet/GetGasLimit.cs:13)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <b6c5d0f08ddf477186f780f49bb14dee>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

Contract call
```
RpcResponseException: intrinsic gas too low: eth_sendRawTransaction
Nethereum.JsonRpc.Client.ClientBase.HandleRpcError (Nethereum.JsonRpc.Client.RpcMessages.RpcResponseMessage response, System.String reqMsg) (at <66374d5b2f39468faa1075c894ca7222>:0)
Nethereum.JsonRpc.Client.ClientBase.SendInnerRequestAsync[T] (Nethereum.JsonRpc.Client.RpcMessages.RpcRequestMessage reqMsg, System.String route) (at <66374d5b2f39468faa1075c894ca7222>:0)
Nethereum.JsonRpc.Client.ClientBase.SendRequestAsync[T] (Nethereum.JsonRpc.Client.RpcRequest request, System.String route) (at <66374d5b2f39468faa1075c894ca7222>:0)
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <9868b380aaa64a8fa33572d5028fb435>:0)
Rethrow as Web3Exception: intrinsic gas too low: eth_sendRawTransaction
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <9868b380aaa64a8fa33572d5028fb435>:0)
Web3Unity.Scripts.Library.Ethers.Contracts.Contract.Send (System.String method, System.Object[] parameters, Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest overwrite) (at <5ae973ee2bcf42eb922496935c4c57aa>:0)
ContractCall.Start () (at Assets/Scripts/Samples/Wallet/ContractCall.cs:12)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <b6c5d0f08ddf477186f780f49bb14dee>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```
Get Nonce **and** Get Tx Status
```
Exception: Invalid account used signing
Nethereum.Web3.Accounts.AccountSignerTransactionManager.SignTransactionRetrievingNextNonceAsync (Nethereum.RPC.Eth.DTOs.TransactionInput transaction) (at <e4c0d072d8984737a021d923143c8fd5>:0)
Nethereum.Web3.Accounts.AccountSignerTransactionManager.SignTransactionAsync (Nethereum.RPC.Eth.DTOs.TransactionInput transaction) (at <e4c0d072d8984737a021d923143c8fd5>:0)
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <9868b380aaa64a8fa33572d5028fb435>:0)
Rethrow as Web3Exception: Invalid account used signing
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <9868b380aaa64a8fa33572d5028fb435>:0)
GetNonce.Start () (at Assets/Scripts/Samples/Wallet/GetNonce.cs:11)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <b6c5d0f08ddf477186f780f49bb14dee>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

**SignVerify** (Sign Message works)
```
Message was signed by a different wallet
UnityEngine.Debug:LogError (object)
SignVerify/<Start>d__0:MoveNext () (at Assets/Scripts/Samples/Wallet/SignVerify.cs:21)
System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<SignVerify/<Start>d__0> (SignVerify/<Start>d__0&)
SignVerify:Start ()
```

**Comment by Arshia001** [comment]
So....

intrinsic gas too low ->
> it is able to tell that the transaction will fail because you haven't given the transaction enough gas to execute with.

32000 execution reverted -> I guess there is something wrong with the contract we're calling, maybe a bad contract address or something?

Both issues sound like we're supplying the wrong parameters to the actual RPC calls. The other two I'll look into.

**Comment by juans-chainsafe** [comment]
> So....
> 
> intrinsic gas too low ->
> 
> > it is able to tell that the transaction will fail because you haven't given the transaction enough gas to execute with.
> 
> 32000 execution reverted -> I guess there is something wrong with the contract we're calling, maybe a bad contract address or something?
> 
> Both issues sound like we're supplying the wrong parameters to the actual RPC calls. The other two I'll look into.

Get Gas Limit I think you are right, but Contract Call works on Remote Wallet (metamask)

**Comment by Arshia001** [comment]
Will look at it then!

**Comment by Arshia001** [comment]
@juans-chainsafe `SignVerify` and `GetNonce` should be good to go.

**Comment by Arshia001** [comment]
@juans-chainsafe about `GetGasLimit`, there is no dependency on the wallet at all, just an RPC call. I just tried it and it works on my end.

**Comment by Arshia001** [comment]
Lastly, about `ContractCall`, I believe this is happening because external wallets supply values for gas limits when it's missing from the transaction request. This does not happen with our in-process wallet, so when the user does not specify gas limits (as is the case with `ContractCall`) the call fails.

**Comment by juans-chainsafe** [comment]
> Lastly, about `ContractCall`, I believe this is happening because external wallets supply values for gas limits when it's missing from the transaction request. This does not happen with our in-process wallet, so when the user does not specify gas limits (as is the case with `ContractCall`) the call fails.

Get Nonce, Get TX status and SignVerify are working now!
@kalambet we need to see what we do with ContractCall then, if we try to fix it or if it will not be supported in Web3Auth

File: src/ChainSafe.GamingSDK.EVM/Web3/Core/Evm/ISigner.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@ public interface ISigner
 
         Task<string> SignTypedData(Domain domain, Dictionary<string, MemberDescription[]> types, MemberValue[] message);
 
+        // TODO: is this the right thing to do?
         Task<string> SignMessage(byte[] message) => SignMessage(message.ToHex());
```
Comment (RyRy79261) [inline src/ChainSafe.GamingSDK.EVM/Web3/Core/Evm/ISigner.cs:17]: I think messages can be any arbitrary string, not just hex

File: src/ChainSafe.GamingSdk.EVM.InProcessSigner/InProcessSigner.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        public Task<string> SignMessage(string message) =>
+            Task.FromResult(messageSigner.EncodeUTF8AndSign(message, privateKey));
+
+        // TODO: test this with Gelato
+        public Task<string> SignTypedData(Domain domain, Dictionary<string, MemberDescription[]> types, MemberValue[] message) =>
+            Task.FromResult(
+                Eip712TypedDataSigner.Current.SignTypedData(
+                    new TypedData<Domain>
+                    {
+                        // TODO: missing the PrimaryType property, not sure if it's required
```
Comment (RyRy79261) [inline src/ChainSafe.GamingSdk.EVM.InProcessSigner/InProcessSigner.cs:36]: Nethereum deduces primary type automatically

**Comment by kalambet** [comment]
@juans-chainsafe @oleksandrchainsafe should be fixed now. Please re-review this one 🫠

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
We are almost done! a lot of improvements were made so thats nice.

I still need to test this with WebGL builds, will take time.

In the mean time, the only issue that I found is with **GetListedCollections** and **ListCollections**. The issue happens when you click "Buy" or "Sell" button in one of the NFT of the collection (slug "soccer2" can be used for testing).

**Logs**
Screenshot 2023-08-09 at 13 34 37
```
RpcResponseException: intrinsic gas too low: eth_sendRawTransaction
Nethereum.JsonRpc.Client.ClientBase.HandleRpcError (Nethereum.JsonRpc.Client.RpcMessages.RpcResponseMessage response, System.String reqMsg) (at <66374d5b2f39468faa1075c894ca7222>:0)
Nethereum.JsonRpc.Client.ClientBase.SendInnerRequestAsync[T] (Nethereum.JsonRpc.Client.RpcMessages.RpcRequestMessage reqMsg, System.String route) (at <66374d5b2f39468faa1075c894ca7222>:0)
Nethereum.JsonRpc.Client.ClientBase.SendRequestAsync[T] (Nethereum.JsonRpc.Client.RpcRequest request, System.String route) (at <66374d5b2f39468faa1075c894ca7222>:0)
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <a3bb78db1e564f62b47bba627635b1c9>:0)
Rethrow as Web3Exception: intrinsic gas too low: eth_sendRawTransaction
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.InProcessTransactionExecutor.SendTransaction (Web3Unity.Scripts.Library.Ethers.Transactions.TransactionRequest transaction) (at <a3bb78db1e564f62b47bba627635b1c9>:0)
GetListedCollections.BuyNFT (System.Int32 nftNumber) (at Assets/Scripts/Prefabs/Minter/GetListedCollections.cs:175)
UnityEngine.Debug:LogException(Exception, Object)
<BuyNFT>d__16:MoveNext() (at Assets/Scripts/Prefabs/Minter/GetListedCollections.cs:180)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
ChainSafe.GamingSdk.EVM.InProcessTransactionExecutor.<SendTransaction>d__4:MoveNext()
UnityEngine.UnitySynchronizationContext:ExecuteTasks() (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

Prefabs used for testing, both are located in Assets/Prefabs/Minter:

- GetListedCollections
- ListCollections

**Review by oleksandrchainsafe — APPROVED** [review]
Looks good to me

**Review by oleksandrchainsafe — DISMISSED** [review]

**Comment by kalambet** [comment]
@juans-chainsafe WebGL and `SendTransaction` fixed!

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
@kalambet GetListedCollections and ListCollections are working now! (hard to test them end to end because we don't have a marketplace UI anymore, so I couldn't find an NFT to buy so I always get an error of insufficient funds, but looks good)

WebGL error was fixed but now we have another issue, if I try to login using web3auth, it all happens in the same window so looks like it can't redirect correctly after the login to the sample scene. Recording:
https://github.com/ChainSafe/web3.unity/assets/104365141/8e7ebea4-523b-4f19-a720-9920a5bc1b40
cc @oleksandrchainsafe

**Comment by juans-chainsafe** [comment]
Also @kalambet me have a new issue in WebGL builds, if you login with Metamask (Remote wallet option) and in your SampleScene have something that needs to sign with Metamask, for example **Transfer20.cs**, I'm getting this error:
```
index-d05f6a39.js:403 Uncaught (in promise) TypeError: invalid BigNumberish string: empty string (argument="request.value", value="", code=INVALID_ARGUMENT, version=6.6.2)
    at oi (index-d05f6a39.js:403:1993)
    at nt (index-d05f6a39.js:403:2185)
    at ye (index-d05f6a39.js:403:2216)
    at St (index-d05f6a39.js:403:5313)
    at yv (index-d05f6a39.js:403:161267)
    at yie._getTransactionRequest (index-d05f6a39.js:403:206664)
    at yie.estimateGas (index-d05f6a39.js:403:207839)
    at index-d05f6a39.js:403:226285
    at $2.sendUncheckedTransaction (index-d05f6a39.js:403:226324)
    at $2.sendTransaction (index-d05f6a39.js:403:226617)
    at async Object.Sie [as sendTransactionData] (index-d05f6a39.js:403:240647)
oi @ index-d05f6a39.js:403
nt @ index-d05f6a39.js:403
ye @ index-d05f6a39.js:403
St @ index-d05f6a39.js:403
yv @ index-d05f6a39.js:403
_getTransactionRequest @ index-d05f6a39.js:403
estimateGas @ index-d05f6a39.js:403
(anonymous) @ index-d05f6a39.js:403
sendUncheckedTransaction @ index-d05f6a39.js:403
sendTransaction @ index-d05f6a39.js:403
await in sendTransaction (async)
_JS_sendTransactionData @ webgl2.framework.js.br:3
$func92760 @ webgl2.wasm.br:0x188bfa9
$_i @ webgl2.wasm.br:0x1a54b38
```

It works on Unity editor and I remember that was working on Gelato PR in webgl

**Review by juans-chainsafe — APPROVED** [review]

**Pull request merged**